[docs] Clarify local/native modules, options for adding custom code (#26602)
expo/expo: docs/pages/workflow/customizing.mdx

import { Terminal } from '~/ui/components/Snippet';

> Just as Expo projects can use third-party React Native packages containing native code, Expo projects are also ready for you to add your own custom native code to your project. Once you create your first [development build](/develop/development-builds/introduction/), you can use any native library or custom native code in your local development environment, as well as in your production app. Development builds work just like Expo Go, allowing you to quickly run your app from the [Expo CLI](/more/expo-cli/), except they include all your specific native dependencies and customizations, and not just what is included with the latest [Expo SDK](/versions/latest/).

Reviewer: ```suggestion
Expo projects can use both third-party React Native libraries with native code and your own custom native code. Creating a [development build](/workflow/overview/#development-builds) allows you to include your specific native dependencies and customizations and not what's in the latest [Expo SDK](/versions/latest/). These native customizations can be any React Native library or your own custom native code.
```

- Let's prefer using library instead of packages
- Avoid using just
- Readability score was a bit high, not easy to skim so rephrased the text for conciseness
- I feel like we don't need to define what development builds does and instead linked to this: https://docs.expo.dev/workflow/overview/#development-builds
- Do we really need to mention Expo Go here?

> For most third-party native modules, thanks to [autolinking](/modules/autolinking/), all you need to do is install the package and create a new development build in order to start using the module's functionality. Some modules will also require you setup a [config-plugin](/config-plugins/introduction) to take care of other native customizations. For writing your own custom native code, there are different options depending on how you want to interact with native code in general.

Reviewer: ```suggestion
For most third-party native libraries, [autolinking](/modules/autolinking/) makes it easy. All you have to do is install the package and create a new development build to use the module. Some modules might also need a [config-plugin](/config-plugins/introduction) for additional native customizations. When adding your own custom native code, the approach varies based on how you want to work with native code.
```

- Avoid using words like thanks to
- Make it concise
- Fix CI warning

![CleanShot 2024-01-24 at 13 44 29@2x](https://github.com/expo/expo/assets/10234615/4b3c6e3d-20c9-46ee-bf05-f38721dd5d3e)

## Create a local module with the Expo Modules API

> By default, Expo projects use [Continuous Native Generation (CNG)](/workflow/continuous-native-generation). This means that projects do not have **android** and **ios** folders containing the native code and configuration used to build the app. Instead, the native projects are generated just before they are built based on your [Expo Config](/workflow/configuration) native modules in **package.json**. This simplifies the management and upgrading of native code, but means you cannot edit the native code directly.

Reviewer: ```suggestion
By default, Expo projects use [Continuous Native Generation (CNG)](/workflow/continuous-native-generation). This means that projects do not have **android** and **ios** directories containing the native code and configuration. Instead, the native directories are generated right before building the app and is based on your [app config](/workflow/configuration) and native modules in **package.json**. This simplifies the management and upgrading of native code.
```

> However, you can still add custom native code to a project using CNG by [creating a local Expo module](modules/get-started/#creating-the-local-expo-module). Local Expo Modules are just like [Expo Modules](/modules/overview/) created by package developers and those used within the Expo SDK itself, but they are not published to npm. Instead, they are created directly inside of your project. This allows you to write custom Swift and Kotlin code that you can interact with from your JavaScript coe, and is included in your development builds and production apps automatically when you build with [EAS Build](/build/introduction) or when you build locally with `npx expo run:[android|iOS]`.

Reviewer: ```suggestion
It might seem you cannot add custom native code to a project using CNG. However, you can still add that by [creating a local Expo module](modules/get-started/#creating-the-local-expo-module). 

Local Expo Modules function similarly to [Expo Modules](/modules/overview/) used by library developers and within the Expo SDK. They are not published to npm. Instead, you create them directly inside your project. 

Creating a local module allows you to write custom Swift and Kotlin code that you can interact with from your JavaScript code. It is also included in the development builds and production apps automatically when you build with [EAS Build](/build/introduction) or when you [compile your app locally](/guides/local-app-development/#local-app-compilation).
```

- I tried to split into multiple paragraph for easier readability. More often, users want to skim, and I think we should allow them to do so by structuring the context in similar ways.
- Also added link to local app compilation as often people ask how that works and what needs to be done, and removed the `npx expo run...` commmand

> ## Manage your own native projects

Reviewer: ```suggestion
## Manage custom native projects
```

I'd avoid using "your own" in section titles. We include section titles in search and "your" is not a phrase that is used by someone using search.

> If you want to opt out of CNG and instead directly manage the code and configuration inside your Android and iOS native projects, you can generate native directories by running `npx expo prebuild` or `npx expo run:[android|iOS]` (which will run `prebuild` automatically). You can also use development builds in this context &mdash; run `npx expo install expo-dev-client` before `prebuild` or `run`.

Reviewer: ```suggestion
You can opt out of CNG by directly managing code and configuration inside **android** and **ios** directories. To generate these directories, run `npx expo prebuild` or [compile your app locally](/guides/local-app-development/#local-app-compilation) (`npx expo run android` or `npx expo run ios`). You can also create a development build before compiling your app locally. To do so, run `npx expo install expo-dev-client` before `prebuild` or `run`.
```

Author: I generally like this, but feel like "by" implies a causal relationship that isn't quite there (there's no native projects to manage until you do something else to opt out of CNG), so accepting and changing a couple of words after :-)

  cmdCopy="npm i -g eas-cli && eas build -p all"
/>

## Create reusable native modules

> As well as being used for adding custom code to single projects, the [Expo Modules API](/modules/overview) enables developers to build reusable modules for Expo and React Native projects using Swift, Kotlin, and TypeScript. These modules are often published to npm; they can also be included as separate packages within a [monorepo](/guides/monorepo). We use the Expo Modules API for most modules in the [Expo SDK](/versions/latest). Learn more about the [Expo Module API design considerations](/modules/overview).

Reviewer: ```suggestion
Apart from adding custom native code to a project, the [Expo Modules API](/modules/overview) enables developers to build reusable modules for Expo and React Native projects using Swift, Kotlin, and TypeScript. These modules are often published to npm. They can also be included as separate packages within a [monorepo](/guides/monorepo). We use the Expo Modules API for most modules in the [Expo SDK](/versions/latest).
```

- I think linking to Expo Modules API overview page once from this is enough
- Also split the sentence instead of using `;`
